### PR TITLE
fix: Do not cache updatable connections

### LINKED ISSUE #15597: [Bug]: API endpoint returning readAddress(..) failed: Connection reset by peer;
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

When a user sets up an Authenticated Rest API Endpoint and runs the query, they get the following error: `readAddress(..) failed: Connection reset by peer; nested exception is io.netty.channel.unix.Errors$NativeIoException: readAddress(..) failed: Connection reset by peer`
unknown 2
unknown 3

### Steps To Reproduce

1. Create an Authenticated Rest API endpoint with the below configuration:

unknown 4
unknown 5

2. Set the content type to `application/json` and body data to
unknown 2

3. Run the API.

### Public Sample App

_No response_

### Version

1.7.10

### PR DESCRIPTION
**Pull request by nidhi-nair**
## Description

We were not refreshing the datasourceContext cache in case of stale connections, this PR fixes that and tests it via `UpdatableConnection` types which would be the most frequent use case that saw such failures. 

Reproduction on release:
1. Create an Authorization Code OAuth2 enabled REST API datasource
2. Go through authentication, create an API and ensure that it works
3. Create a new workspace, and fork this app into that workspace
4. Try to authorize the datasource in the new workspace and re-run the API. This will not work on release. It should work on this branch,

Fixes #15597 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
-JUnit test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceContextServiceTest.java [hunk 1]
```diff
@@ -159,22 +163,78 @@ public void testCachedDatasourceCreate() {
...
+                    DatasourceContext<?> dsContext2 = tuple.getT2();
                     /* They can only be equal if the `datasourceCreate` method was called only once */
                     assertEquals(dsContext1.getConnection(), dsContext2.getConnection());
                     assertEquals("connection_1", dsContext1.getConnection());
                 })
                 .verifyComplete();
     }
+
+    /**
+     * This test checks that if `getCachedDatasourceCreate` method is called two times for the same datasource id, then
```
Comment (sumitsum) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceContextServiceTest.java:187]: Minor: please update comment.

**Review by sumitsum — CHANGES_REQUESTED** [review]

**Comment by sumitsum** [comment]
@nidhi-nair can you please create a separate issue to add a cypress test to check for refresh token OAuth flow since now QA team has added some automation capability around OAuth.

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceContextServiceTest.java (hunk 1 above)
Comment (nidhi-nair) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceContextServiceTest.java:187]: 🤔 I actually did, I replaced the end state only with Updatable connection type :P

Comment (sumitsum) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/services/DatasourceContextServiceTest.java:187]: Oh ... yup my mistake ....:)

**Review by sumitsum — DISMISSED** [review]

**Pull request merged**